File: src/DesignPatterns/src/Java/002-GOF-Patterns/002-StructuralPatterns/004-Proxy/ImageProxy/src/org/csystem/app/BitmapImage.java
package org.csystem.app;

public class BitmapImage implements IImage {
    private String m_filename;

    private void loadImage()
    {
        System.out.println("Image loaded");
    }

    public BitmapImage(String filename)
    {
        m_filename = filename;
        loadImage();
    }

    @Override
    public void show()
    {
        System.out.println("Image display");
    }
}

File: src/DesignPatterns/src/C++/002-GOFPatterns/002-StructuralPatterns/004-Proxy/BitmapImageProxy/BitmapImage.cpp
#include "BitmapImage.h"

#include <iostream>

using namespace std;

BitmapImage::BitmapImage(const char *url)
{
	loadImage();
	std::strcpy(m_url, url);
}

void BitmapImage::Show()
{
	cout << "Show Image:" << m_url << endl;
}

void BitmapImage::SetUrl(const char *url)
{
	cout << "Set url" << endl;
	strcpy(m_url, url);
	loadImage();
}

void BitmapImage::loadImage()
{
	cout << "loading image" << endl;
}
